<a href="https://colab.research.google.com/github/MariadelMar04/Analitica-3/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Arboles de Decisión / Reto 2**

**Tech Sales Report**

**María del Mar Muñoz Lopez**

**Descripción de las variables (según libro guía)**

**Business:** Uno de los dos grupos de productos: Hardware y Software

**Age:** Edad del empleado

**Female:** 1 – Mujer – 0 Caso contrario

**Years:** el número de años que el empleado ha estado en la empresa

**College:** Si el empleado tiene o no un título universitario de cuatro años (Sí/No)

**Certificates:** El número de certifi caciones profesionales relevantes que ha obtenido cada empleado

**Feedback:** El puntaje de retroalimentación promedio que cada empleado recibede sus compañeros y supervisor en la evaluación anual de 360 grados. Las puntuaciones posibles van de 0 (la más baja) a 4 (la más alta)

**Salary:** Sueldo base anual de cada empleado.

**NPS:** El puntaje neto del promotor (NPS) es un indicador clave de la satisfacción y lealtad del cliente


0. Cargar la libreria de trabajo

In [ ]:
import numpy as np
import pandas as pd

#Para implementación y evaluación del modelo
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

#Para la carga automatica de los archivos desde el drive
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Carga base de datos

In [ ]:
nxl='/content/drive/MyDrive/2. TechSalesReport_Int_L.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.head(100) #Solo muestra 100 datos

# Convertir Business a valores numericos
XDB['Business']=pd.Categorical(XDB['Business']).codes
XDB=XDB.dropna()
print(XDB)


       Sales_Rep  Business  Age  Female  Years College Personality  \
0              1         0   59       1      2     Yes    Diplomat   
1              2         0   52       0     10     Yes    Diplomat   
2              3         1   47       1      1     Yes    Explorer   
3              4         0   61       0      2     Yes    Diplomat   
4              5         1   39       0      1      No    Diplomat   
...          ...       ...  ...     ...    ...     ...         ...   
21985      21986         1   44       0      1     Yes    Diplomat   
21986      21987         1   23       1      6     Yes     Analyst   
21987      21988         0   48       1      4     Yes    Sentinel   
21988      21989         1   29       0      4     Yes     Analyst   
21989      21990         1   23       1      2     Yes    Explorer   

       Certificates  Feedback  Salary  NPS  
0                 1      2.01   70200    5  
1                 4      3.64  133000   10  
2                 1     

In [ ]:
#Selecionar Variables
XD=XDB[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']]
print(XD)
yd=XDB[['College']]
print(yd)

       Business  Age  Female  Years  Certificates  Feedback  Salary  NPS
0             0   59       1      2             1      2.01   70200    5
1             0   52       0     10             4      3.64  133000   10
2             1   47       1      1             1      3.88   52600    8
3             0   61       0      2             3      2.70   96000    6
4             1   39       0      1             5      3.44  122000    7
...         ...  ...     ...    ...           ...       ...     ...  ...
21985         1   44       0      1             4      1.80  128000    7
21986         1   23       1      6             6      1.77   53400    8
21987         0   48       1      4             0      2.46   50400    4
21988         1   29       0      4             2      3.68   70000   10
21989         1   23       1      2             1      2.13   47400    5

[21990 rows x 8 columns]
      College
0         Yes
1         Yes
2         Yes
3         Yes
4          No
...       ...


2. Implementación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)

#Max_depth = numero de variables disponibles
#Gini es el criterio que indica el grado de error del modelo, en la clasificación de solicitantes
mar.fit(XD,yd)

#Se procede con el despliegue del arbol
from sklearn.tree import export_graphviz #Exportar
from pydotplus import graph_from_dot_data #Crear grafico
vs=['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('arbol.png')

XD #Para ver tabla


Business  Age  Female  Years  Certificates  Feedback  Salary  NPS
0             0   59       1      2             1      2.01   70200    5
1             0   52       0     10             4      3.64  133000   10
2             1   47       1      1             1      3.88   52600    8
3             0   61       0      2             3      2.70   96000    6
4             1   39       0      1             5      3.44  122000    7
...         ...  ...     ...    ...           ...       ...     ...  ...
21985         1   44       0      1             4      1.80  128000    7
21986         1   23       1      6             6      1.77   53400    8
21987         0   48       1      4             0      2.46   50400    4
21988         1   29       0      4             2      3.68   70000   10
21989         1   23       1      2             1      2.13   47400    5

[21990 rows x 8 columns]

3. Desempeño del modelo

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VN+VP)/(VN+FP+FN+VP)
print("Exactitud del modelo:",Ex)

TE=(FN+FP)/(VN+FP+FN+VP)
print("Tasa de Error:",TE)

Sen=VP/(VP+FN)
print("Sensibilidad del modelo:",Sen)

Sp=VN/(VN+FP)
print("Especificidad del modelo:",Sp)

pre=VP/(VP+FP)
print("Precisión del modelo:",pre)

pren=VN/(VN+FN)
print("Precisión Negativa del modelo:",pren)



[[  114  4356]
 [   69 17451]]
Exactitud del modelo: 0.7987721691678036
Tasa de Error: 0.20122783083219645
Sensibilidad del modelo: 0.9960616438356165
Especificidad del modelo: 0.025503355704697986
Precisión del modelo: 0.8002476269087908
Precisión Negativa del modelo: 0.6229508196721312


In [ ]:
# Implementar modelo Naive Bayes

from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd.values.ravel())

GaussianNB()

In [ ]:
from time import ctime
#Se obtiene la información del modelo
ct=mnb.class_count_
print("¿Como estan separados los datos?\n",ct)
print("El total de los datos es:\n",len(XD))

¿Como estan separados los datos?
 [ 4470. 17520.]
El total de los datos es:
 21990


5. Llevar a cabo el pronóstico del título Universitario para las personas agrupadas en la base de datos de pronóstico.

In [ ]:
import pandas as pd
# Instala la librería openpyxl si no la tienes:
# !pip install openpyxl

# Crear un DataFrame con los pronósticos
# Suponiendo que 'ydp' contiene los pronósticos del modelo
# y 'XD' contiene los datos de entrada
df_pronostico = pd.DataFrame({'Pronostico': ydp}, index=XD.index)

# Unir el DataFrame de pronósticos con los datos originales (opcional)
df_completo = pd.concat([XD, df_pronostico], axis=1)

# Guardar el DataFrame en un archivo de Excel
df_completo.to_excel('informe_pronostico.xlsx', sheet_name='Pronosticos', index=False)

## **Análisis de Resultados**

Para la implementación del modelo de Árbol de Decisión, se trabajó con un conjunto de datos que incluyó un total de 21,990 registros. Dentro de estos, 4,470 corresponden a personas no inscritas y 17,520 a personas inscritas. Se emplearon 8 variables de entrada, las cuales incluyen: Businesss, age, Female, Years, College, Certificates, Feedback, Salary y NPS. Y la variable de salida utilizada para la clasificación fue College.

Para garantizar la calidad del análisis, se realizó un preprocesamiento de los datos, asegurando la limpieza y coherencia de la información antes de la modelización. Luego de la ejecución del modelo, se obtuvo una exactitud del 79.88%, lo que indica un buen nivel de precisión en la clasificación de los registros. La tasa de error del modelo se situó en 20.12%, reflejando la proporción de casos en los que la predicción fue incorrecta. Un aspecto relevante a destacar es la alta sensibilidad del modelo, alcanzando un 99.60%, lo que sugiere que el modelo es sumamente eficaz en la identificación de los casos positivos. Sin embargo, la especificidad se ubicó en un bajo 2.55%, lo que indica dificultades en la correcta clasificación de los casos negativos. La precisión del modelo fue del 80.02%, mientras que la precisión negativa se situó en 62.29%, mostrando un desempeño aceptable en términos generales.

Uno de los aspectos más notables de la estructura del árbol de decisión generado es que no presenta ningún nodo puro.  Si bien el modelo logró una clasificación adecuada con una alta sensibilidad, es recomendable explorar técnicas adicionales de optimización para mejorar la especificidad y balancear mejor la clasificación de las distintas categorías